B024 false negative when there's a class var

```python
from abc import ABC


class Foo(ABC):
    BAR = 1

    def method(self):
        bar()
```

Because of the `BAR = 1` declaration, B024 is considering `Foo` an ABC, but it isn't.

Expected:

```
$ flake8 --isolated --select B024 abc.py
abc.py:4:1: B024 Foo is an abstract base class, but none of the methods it defines are abstract. This is not necessarily an error, but you might have forgotten to add the @abstractmethod decorator, potentially in conjunction with @classmethod, @property and/or @staticmethod.
$
```

But got:

```
$ flake8 --isolated --select B024 abc.py
$
```

Info:

```
$ flake8 --version
7.0.0 (flake8-bugbear: 24.4.26, mccabe: 0.7.0, pycodestyle: 2.11.1, pyflakes: 3.2.0) CPython 3.12.2 on Darwin
```

This looks related to the fix for https://github.com/PyCQA/flake8-bugbear/issues/293.

If I comment out this code, it works as expected:

https://github.com/PyCQA/flake8-bugbear/blob/d1aec4cbef7c4a49147c428b7e4a97e497b5d163/bugbear.py#L925-L929

By the way, this same issue was reported to ruff in https://github.com/astral-sh/ruff/issues/11208 and they got a fix. Maybe here a similar patch would work?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -9,7 +9,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12", "3.13-dev"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12", "3.13"]
 
         os: [ubuntu-latest]
 

--- a/README.rst
+++ b/README.rst
@@ -357,6 +357,11 @@ MIT
 Change Log
 ----------
 
+
+FUTURE
+~~~~~~
+* B024: No longer treats assigned class variables as abstract (#471)
+
 24.8.19
 ~~~~~~~
 

--- a/bugbear.py
+++ b/bugbear.py
@@ -996,7 +996,7 @@ class BugBearVisitor(ast.NodeVisitor):
         for stmt in node.body:
             # https://github.com/PyCQA/flake8-bugbear/issues/293
             # Ignore abc's that declares a class attribute that must be set
-            if isinstance(stmt, (ast.AnnAssign, ast.Assign)):
+            if isinstance(stmt, ast.AnnAssign) and stmt.value is None:
                 has_abstract_method = True
                 continue
 

--- a/tests/b024.py
+++ b/tests/b024.py
@@ -10,7 +10,7 @@ import foo
 
 """
 Should emit:
-B024 - on lines 17, 34, 52, 58, 69, 74, 84, 89
+B024 - on lines 17, 52, 58, 69, 74, 123, 129
 """
 
 
@@ -31,7 +31,7 @@ class Base_3(ABC):
         foo()
 
 
-class Base_4(ABC):
+class Base_4(ABC): # safe
     @notabc.abstractmethod
     def method(self):
         foo()
@@ -112,16 +112,24 @@ class keyword_abc_2(metaclass=abc.ABC):  # safe
         foo()
 
 
-class abc_set_class_variable_1(ABC):  # safe
+# safe, see https://github.com/PyCQA/flake8-bugbear/issues/293
+class abc_annasign_empty_class_variable_1(ABC):
     foo: int
+    def method(self):
+        foo()
 
 
-class abc_set_class_variable_2(ABC):  # safe
+# *not* safe, see https://github.com/PyCQA/flake8-bugbear/issues/471
+class abc_assign_class_variable(ABC):
     foo = 2
+    def method(self):
+        foo()
 
 
-class abc_set_class_variable_3(ABC):  # safe
+class abc_annassign_class_variable(ABC):  # *not* safe, see #471
     foo: int = 2
+    def method(self):
+        foo()
 
 
 # this doesn't actually declare a class variable, it's just an expression

--- a/tests/test_bugbear.py
+++ b/tests/test_bugbear.py
@@ -461,6 +461,8 @@ class BugbearTestCase(unittest.TestCase):
             B024(58, 0, vars=("MetaBase_1",)),
             B024(69, 0, vars=("abc_Base_1",)),
             B024(74, 0, vars=("abc_Base_2",)),
+            B024(123, 0, vars=("abc_assign_class_variable",)),
+            B024(129, 0, vars=("abc_annassign_class_variable",)),
         )
         self.assertEqual(errors, expected)
 

